### DISCUSSION TITLE
Binding not updated when updating Collection  using Dispatcher

### DISCUSSION THREAD

**Discussion by julian-baumann** [AvaloniaUI/Avalonia#17084] [Q&A]
### Describe the bug

I am using the DataGrid component and Avalonia Version 11.1.3. Using a ViewModel I defined a Property `Entries` where I bind the `ItemsSource` to. I am calling a method using a Command, that updates this collection. When doing it in a synchronous environment everything works just fine, but as soon as I used `Task.Run` to run the logic and updated it using `Dispatcher.UIThread.Invoke`, nothing happens. No exception is thrown, and no UI updates in the DataGrid.

```cs
using ReactiveUI.SourceGenerators;

[Reactive] private ObservableCollection<Dictionary<string, string>> _entries = [];

public void GetEntries()
{
    Task.Run(() =>
    {
        Dispatcher.UIThread.Invoke(() =>
        {
            Entries = [];
        });

        ... Doing some stuff

        Dispatcher.UIThread.Invoke(() =>
        {
            Entries.Add(properties);
        });
    });
}
```

### To Reproduce

Using DataGrid, Bindings and a ReactiveUI ViewModel

### Avalonia version

11.1.3

### OS

macOS

**Comment by timunie** [comment]
Dispatcher.UiTread.Post should be used iirc. Can you try this?

**Comment by timunie** [comment]
One more hint, you can try DynamicData if you need async manipulations. It's shipped with ReactiveUI anyways

**Comment by julian-baumann** [comment]
Both approaches don't work. Neither `Dispatcher.UiTread.Post` nor using `SourceList` with an `ReadOnlyObservableCollection` updates the UI

**Comment by timunie** [comment]
```cs
    [RelayCommand]
    private async Task LoadPeopleAsync()
    {
        await Task.Delay(1000);
        People = new ObservableCollection<Person>();
        
        await Task.Delay(1000);
        
        People.Add(new Person { FirstName = "John", LastName = "Doe" });
        People.Add(new Person { FirstName = "Jane", LastName = "Doe" });
        await Task.Delay(1000);
        People.Add(new Person { FirstName = "Julie", LastName = "Doe" });

        Dispatcher.UIThread.Invoke(() =>
        {
            People.Add(new Person { FirstName = "Norman", LastName = "Doe" });
        });
    }
```

Just tried it on my own and all of the above works. Note: I use a Command here to bind it to a button, but this is not required.

I see that you don't even await your Task, so maybe this is the issue. In any case, I think it is something in your code that blocks the UI: So I make it a Q&A for now.

**Comment by julian-baumann** [comment]
Still no success. Awaiting the Task also doesn't change the outcome. Also not using the Dispatcher results in a ` Call from invalid thread` exception. Why is it working in your example?

**Comment by julian-baumann** [comment]
Okay, so It seems to be working when I do this:
```cs
public async Task GetEntries()
{
    await Task.Run(() =>
    {
        var localEntries = [];
        // ... Doing some stuff

        
        Entries = localEntries;
    });
}
```

Fells like using `Add` to dynamically update the ObservableCollection `Entries` does not trigger a UI update, but reassigning the Property does.

**Reply by stevemonaco** [reply]
`Entries.Add` should update the UI. If you're mutating a dictionary already stored within the `ObservableCollection<T>`, then that won't update anything. These changes should be dispatched because they're triggering UI updates, so the `Task.Run` is probably pointless.

**Reply by timunie** [reply]
@julian-baumann can you attach a minimal repo? I think you are doing something strange / risky. Easiest if you upload it to GitHub and share the link with us.

**Reply by julian-baumann** [reply]
https://github.com/cyancor/litedb-explorer

Relevant parts
The `DataGrid` in [MainWindow.xaml](https://github.com/cyancor/litedb-explorer/blob/master/Sources/LiteDB.Explorer/Views/MainWindow.axaml)
And `GetEntries()` in [MainWindowViewModel.cs](https://github.com/cyancor/litedb-explorer/blob/master/Sources/LiteDB.Explorer/ViewModels/MainWindowViewModel.cs)

Run the Program click "Open new Database" and select this file [Resources/TestDatabase.db](https://github.com/cyancor/litedb-explorer/blob/master/Resources/TestDatabase.db)

**Reply by stevemonaco** [reply]
Code seems to work as-is. I had to disable the following line because appstate.json wasn't included:
```cs
//RxApp.SuspensionHost.SetupDefaultSuspendResume(new NewtonsoftJsonSuspensionDriver("appstate.json"));
```

![litedb-explorer](https://github.com/user-attachments/assets/c05b9671-da27-489f-bc07-f0f0b7f3414e)

The app doesn't seem to close and I need to force quit.

**Reply by julian-baumann** [reply]
Ah yes, sorry. This is the working code. As you may see in Line 42 in [MainWindowViewModel.cs](https://github.com/cyancor/litedb-explorer/blob/master/Sources/LiteDB.Explorer/ViewModels/MainWindowViewModel.cs), I create a temporary entries list, add all entries to that list and then later on assign this list to the `Entries` Property, which is bound to the UI.

What doesn't work is this (found in the `bug/binding` branch):
[MainWindowViewModel.cs](https://github.com/cyancor/litedb-explorer/blob/bug/binding/Sources/LiteDB.Explorer/ViewModels/MainWindowViewModel.cs)
Seems like adding the entries in the for loop directly does not update the UI. Sometimes I get one entry only and then it throws a `Call from invalid Thread` exception. If I wrap this `Add` call in `Dispatcher.UIThread.Post` or `Dispatcher.UIThread.Invoke` it doesn't update the UI at all

**Reply by julian-baumann** [reply]
> Code seems to work as-is. I had to disable the following line because appstate.json wasn't included:

Why did you had to disable this line? Did it crash? I tried it multiple times on mac, linux and windows and it works fine. Should create the file, when closing the program. This may also be the issue you mentioned on why the app doesn't terminate correctly.

**Reply by stevemonaco** [reply]
> Why did you had to disable this line? Did it crash?

Yes, your app doesn't ship appstate.json. Therefore the file isn't not available in a first-run scenario and it doesn't have a fallback.

![image](https://github.com/user-attachments/assets/31d8cbfb-0686-48ba-9125-1f4aca94161a)

**Reply by stevemonaco** [reply]
This might be a bug in DataGrid or some limitation. I'm not able to make the dispatch-per-item work as-is (this isn't a good idea anyways). I'm pretty much required to create new collections. Quickly having at least one item in the binded collection seems incredibly important. Once that happens, I can do any operation as expected, including the dispatch-per-item. Otherwise, I need to be very careful about creating the collection *and* immediately add the first item to it. No await or dispatch in-between.

Also, you should really be awaiting any `async` method you call. Though this wasn't the core problem...at least not yet.

**Reply by IanRawley** [reply]
> > Why did you had to disable this line? Did it crash?
> 
> Yes, your app doesn't ship appstate.json. Therefore the file isn't not available in a first-run scenario and it doesn't have a fallback.
> 
> ![image](https://private-user-images.githubusercontent.com/1782158/370826251-31d8cbfb-0686-48ba-9125-1f4aca94161a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczMjE5MjQsIm5iZiI6MTcyNzMyMTYyNCwicGF0aCI6Ii8xNzgyMTU4LzM3MDgyNjI1MS0zMWQ4Y2JmYi0wNjg2LTQ4YmEtOTEyNS0xZjRhY2E5NDE2MWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjZUMDMzMzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGJiODA0Mjc3NTI2OWM3ZDAwZGIxMWRjMTBhNTdiM2I3YmM4ZWNjOTdmZTA2YjFlYTk0MGY1MDQ2ZTRkZjRmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.MLrUKbXQtftyaz0W305awuy5eQoc4fat3DcKyrYO3Tw)

That's pretty standard in the Serialization tutorials for persisting app state.  I think you'll find if you let it continue the AutoSuspendHelper will catch that exception and everything will actually work, provided the AutoSuspendHelper has been given a way to instantiate a new state object if things break.

**Reply by IanRawley** [reply]
Having said that, calling base.OnFrameworkInitializationComplete() twice looks pretty fishy to me.
```
            base.OnFrameworkInitializationCompleted();


            desktop.MainWindow = new StartWindow
            {
                DataContext = state
            };
        }


        base.OnFrameworkInitializationCompleted();
```

Not actually a problem I don't think, but a bit fishy.
More likely issues are that separate threads aren't dispatching all things that touch the UI to the UI thread, and then I'd guess that not awaiting the dispatch makes it not execute for some reason.

**Reply by IanRawley** [reply]
Hang on.  At least in the OP, you do this: 

```
Task.Run(() =>
    {
        Dispatcher.UIThread.Invoke(() =>
        {
            Entries = [];
        });

        ... Doing some stuff

        Dispatcher.UIThread.Invoke(() =>
        {
            Entries.Add(properties);
        });
    });
```

Given you don't await, there's no guarantee that the first Invoke is actually completed before the second, is there?  So it's quite possible it is working, but adding all the properties to Entries and then setting Entries to an empty collection?  And given that it started empty, that would be why you don't see any changes.

**Reply by stevemonaco** [reply]
I corrected bad async usage in several spots during my troubleshooting and it didn't help. I double-checked the `DataGrid.ItemsSource` in several of the "bad runs" with 0 entries visually displayed while DevTools showed the binded collection had several items. I couldn't locate an exact root cause.

**Reply by timunie** [reply]
The root cause is probably DataGrid that has `AutoGenerateColumns` which is only fired if you set a new `ItemsSource` and while this is empty, no columns are generated 🤷. So either add a dummy item or don't rely on auto generated columns. 

Alternative: Give a name to the datagrid and do this:
```cs
MyDataGrid.Columns.Add([...]); 
```
